pkg/auth/providers/migration/activedirectory_migration.go

```diff
@@ -0,0 +1,243 @@
+		return errors.Wrap(err, "failed updating user object")
+	}
+	return nil
+}
+
+func (m *adMigration) migrateTokens(userName string, newPrincipalID string) error {
+	userTokens, err := m.tokenIndexer.ByIndex(tokenByUserRefKey, userName)
+	if err != nil {
+		return errors.Wrap(err, "failed to fetch tokens")
+	}
+
+	for _, obj := range userTokens {
+		token, ok := obj.(*v3.Token)
+		if !ok {
+			return errors.Errorf("failed to convert object to Token for user %v principalId %v", userName, newPrincipalID)
+		}
+		token.UserPrincipal.Name = newPrincipalID
+		_, e := m.tokens.Update(token)
+		if e != nil {
+			logrus.Errorf("unable to update token %v for principalId %v", token.Name, newPrincipalID)
+		}
+	}
+	return nil
+}
+
+func (m *adMigration) migrateCRTB(newPrincipalID string, dn string) error {
+	userCRTBs, err := m.crtbIndexer.ByIndex(crtbsByPrincipalAndUserIndex, dn)
+	if err != nil {
+		return errors.Wrap(err, "failed to fetch CRTBs")
+	}
+	for _, crtb := range userCRTBs {
+		oldCrtb, ok := crtb.(*v3.ClusterRoleTemplateBinding)
+		if !ok {
+			return errors.Errorf("failed to convert object to CRTB for principalId %v", newPrincipalID)
+		}
+		newCrtb := &v3.ClusterRoleTemplateBinding{
+			ObjectMeta: v1.ObjectMeta{
+				Name:         "",
+				Namespace:    oldCrtb.ObjectMeta.Namespace,
+				GenerateName: "crtb-",
+			},
+			ClusterName:       oldCrtb.ClusterName,
+			UserName:          oldCrtb.UserName,
+			RoleTemplateName:  oldCrtb.RoleTemplateName,
+			UserPrincipalName: newPrincipalID,
+		}
+		_, err := m.crtbs.Create(newCrtb)
+		if err != nil {
+			return errors.Wrap(err, "unable to create new CRTB")
+		}
+		err = m.crtbs.DeleteNamespaced(oldCrtb.Namespace, oldCrtb.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			return errors.Wrap(err, "unable to delete CRTB")
+		}
+	}
+	return nil
+}
+
+func (m *adMigration) migratePRTB(newPrincipalID string, dn string) error {
+	userPRTBs, err := m.prtbIndexer.ByIndex(prtbsByPrincipalAndUserIndex, dn)
+	if err != nil {
+		return errors.Wrap(err, "failed to fetch PRTBs")
+	}
+	for _, prtb := range userPRTBs {
+		oldPrtb, ok := prtb.(*v3.ProjectRoleTemplateBinding)
+		if !ok {
+			return errors.Errorf("failed to convert object to PRTB for principalId %v", newPrincipalID)
+		}
+		newPrtb := &v3.ProjectRoleTemplateBinding{
+			ObjectMeta: v1.ObjectMeta{
+				Name:         "",
+				Namespace:    oldPrtb.ObjectMeta.Namespace,
+				GenerateName: "prtb-",
+			},
+			ProjectName:       oldPrtb.ProjectName,
+			UserName:          oldPrtb.UserName,
+			RoleTemplateName:  oldPrtb.RoleTemplateName,
+			UserPrincipalName: newPrincipalID,
+		}
+		_, err := m.prtbs.Create(newPrtb)
+		if err != nil {
+			return errors.Wrap(err, "unable to create new PRTB")
+		}
+		err = m.prtbs.DeleteNamespaced(oldPrtb.Namespace, oldPrtb.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			return errors.Wrap(err, "unable to delete PRTB")
+		}
+	}
+	return nil
+}
+
+func (m *adMigration) isDNBased(principal string) bool {
```

Comment:
Does this need to be a method? It doesn't use anything from the receiver...

Reply:
You could also simplify this down to a single expression.
`return strings.Contains(principal, ",")`

Reply:
Yeah, I could probably just drop this.  At one point, the check was far more elaborate.